OpenRA.Mods.D2k/Traits/Spicebloom.cs

```diff
@@ -0,0 +1,194 @@
+﻿using OpenRA.GameRules;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Support;
+using OpenRA.Traits;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	public class SpicebloomInfo : ITraitInfo, Requires<RenderSpritesInfo>
+	{
+		public readonly string[] GrowthSequences = { "grow1", "grow2", "grow3" };
+
+		[Desc("The amount of time (in ticks) that the spicebloom will take to respawn.")]
+		public readonly int[] RespawnTime = { 750, 1500 };
+
+		[Desc("The amount of time (in ticks) that the spicebloom will take to grow.")]
+		public readonly int[] GrowthTime = { 750, 1500 };
+
+		public readonly string ResourceType = "Spice";
+
+		[Desc("The weapon to use as spice")]
+		[WeaponReference]
+		public readonly string Weapon = "SpiceExplosion";
+
+		[Desc("The amount of spice to expel.")]
+		public readonly int Pieces = 7;
+
+		[Desc("The maximum distance in cells that spice may be expelled.")]
+		public readonly int Range = 5;
+
+		public object Create(ActorInitializer init)
+		{
+			return new Spicebloom(init, this);
+		}
+	}
+
+	public class Spicebloom : ITick, ICrushable, INotifyDamage
+	{
+		readonly Actor self;
+		readonly SpicebloomInfo info;
+		readonly ResourceType resType;
+		readonly ResourceLayer resLayer;
+		readonly RenderSprites render;
+
+		Animation anim;
+		AnimationWithOffset anim2;
+		int explodedTicks;
+		bool growing;
+		int growTicks;
+		[Sync] int respawnTime;
+		[Sync] int growTime;
+
+		public Spicebloom(ActorInitializer init, SpicebloomInfo info)
+		{
+			this.info = info;
+			self = init.Self;
+
+			resType = self.World.WorldActor.TraitsImplementing<ResourceType>()
+							.FirstOrDefault(t => t.Info.Name == info.ResourceType);
+
+			if (resType == null)
+				throw new InvalidOperationException("No such resource type `{0}`".F(info.ResourceType));
+
+			resLayer = self.World.WorldActor.Trait<ResourceLayer>();
+
+			render = self.Trait<RenderSprites>();
+
+			anim = new Animation(init.Self.World, render.GetImage(self));
+			anim2 = new AnimationWithOffset(anim, null, () => false);
```

Comment:
Instead of adding and removing it, you probably want to tie up the IsVisible func here.